core/src/main/java/org/apache/brooklyn/core/config/internal/AbstractConfigMapImpl.java

```diff
@@ -163,13 +163,18 @@ public ConfigBag getLocalConfigBag() {
         return putAllOwnConfigIntoSafely(ConfigBag.newInstance()).seal();
     }
 
-    public Object setConfig(ConfigKey<?> key, Object v) {
-        Object val = coerceConfigVal(key, v);
+    public Object setConfig(final ConfigKey<?> key, Object v) {
+        // Use our own key for writing, (e.g. in-case it should (or should not) be a structured key like MapConfigKey).
+        // This is same logic as for getConfig, except we only have to look at our own container.
+        ConfigKey<?> ownKey = getKeyAtContainer(getContainer(), key);
+        if (ownKey==null) ownKey = key;
+
+        Object val = coerceConfigVal(ownKey, v);
```

Comment:
I guess it's useful for catching errors early. On the other hand the uncertainty of what config key (and type) will be used to get values vs setting them makes it harder to reason about the behaviour.
Say we are setting a double with an integer key - it will throw away the fractional part. Then getting it back with a `Double` key.